### PR TITLE
Patch Speed Monitor MFU

### PR DESCRIPTION
**Pull request by mvpatel2000**
# What does this PR do?

As title. Fixes speed monitor bug which has a problem with flops/MFU. We now keep a record of flop utilization, as the values might be different depending on the level of padding in the batch. We also correctly sum the flop counts from all the devices

image

# What issue(s) does this change relate to?

[CO-1838](https://mosaicml.atlassian.net/browse/CO-1838)

[CO-1838]: https://mosaicml.atlassian.net/browse/CO-1838?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: composer/callbacks/speed_monitor.py [hunk 1]
```diff
@@ -284,22 +285,37 @@ def batch_end(self, state: State, logger: Logger):
...
+            model_flops_per_batch = composer_model.flops_per_batch  # type: ignore
+            if not isinstance(model_flops_per_batch, Callable):
+                raise TypeError('flops_per_batch must a callable accepting a batch and '
+                                f'returning an int or float. Instead, got {type(model_flops_per_batch)}.')
+            device_flops_per_batch = model_flops_per_batch(state.batch)
+
+            # Sum flops across all ranks since each rank computes the flops for its own batch
+            flops_per_batch_tensor = state.device.tensor_to_device(
+                torch.tensor(device_flops_per_batch, dtype=torch.float))
+            dist.all_reduce(flops_per_batch_tensor, reduce_operation='SUM')
```
Comment (abhi-mosaic) [inline composer/callbacks/speed_monitor.py:302]: I thiiink there might be a `dist.all_reduce_object(..)` that could save you the `torch.tensor` stuff but don't worry about it unless you're confident that it's safe

Comment (mvpatel2000) [reply composer/callbacks/speed_monitor.py:302]: Only have an `all_gather_object` :(

I'm going to leave off for now because I don't want to rerun the tests ahead of release deadline, but this is probably a reasonable update we'd want to do in dist since this tensor wrapping occurs in a few places

Comment (vchiley) [reply composer/callbacks/speed_monitor.py:302]: `flops_per_batch_tensor  = sum(all_gather_object(device_flops_per_batch))`

File: composer/callbacks/speed_monitor.py [hunk 2]
```diff
@@ -284,22 +285,37 @@ def batch_end(self, state: State, logger: Logger):
...
+            flops_per_batch = flops_per_batch_tensor.item()
```
Comment (vchiley) [inline composer/callbacks/speed_monitor.py:303]:
```suggestion
            flops_per_batch  = sum(all_gather_object(device_flops_per_batch))
```

would that work?

**Review by vchiley — APPROVED** [review]
lgtm

(with one nit cmt)
